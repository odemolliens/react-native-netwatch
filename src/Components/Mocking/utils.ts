import RNFS from 'react-native-fs';

export const FILE_PATH = RNFS.DocumentDirectoryPath + 'Netwatch.mockResponses.35435664445664';

export const mockResponses: MockResponse[] = [];

export interface MockResponse {
  url: string;
  method?: string;
  response?: string;
  headers?: string;
  statusCode?: number;
  timeout?: number;
  active?: boolean;
  date?: number;
}

export function getMockResponses() {
  return mockResponses;
}

export function clearMockResponses() {
  mockResponses.length = 0;
  RNFS.unlink(FILE_PATH);
}

export function resetMockResponses(responses: string) {
  clearMockResponses();
  mockResponses.push(...JSON.parse(responses));
  RNFS.writeFile(FILE_PATH, responses, 'utf8');
}

export function mockRequestWithResponse(mockResponse: MockResponse) {
  let found = false;
  for (let i = 0; i < mockResponses.length; i++) {
    if (mockResponses[i].url === mockResponse.url && mockResponses[i].method === mockResponse.method) {
      mockResponses[i] = {
        ...mockResponse,
        active: true,
      };
      found = true;
      break;
    }
  }
  if (!found) {
    mockResponses.push({
      ...mockResponse,
      active: true,
    });
  }
  RNFS.writeFile(FILE_PATH, JSON.stringify(mockResponses), 'utf8');
}

let originalSend: any;

export function extractURL(url: string) {
  const urlParts = url.split('?');
  return urlParts[0];
}

// istanbul ignore next
export function setupMocks() {
  RNFS.readFile(FILE_PATH, 'utf8')
    .then((mockResponsesString: string) => {
      mockResponses.push(...(JSON.parse(mockResponsesString || '[]') as MockResponse[]));
    })
    .catch(() => console.log('No mock responses found'));
  if (originalSend) {
    console.log('mockResponses already setup');
    return;
  }
  originalSend = XMLHttpRequest.prototype.send;
  try {
    XMLHttpRequest.prototype.send = function (body: any) {
      // @ts-ignore
      const mockRequest = mockResponses.find(mock => {
        // @ts-ignore
        try {
          return (
            mock.active &&
            // @ts-ignore
            mock.url === extractURL(this._url) &&
            // @ts-ignore
            (!mock._method || (mock.method && mock._method === this._method))
          );
        } catch (e) {
          return false;
        }
      });

      if (mockRequest) {
        Object.defineProperty(this, 'response', { writable: true });
        Object.defineProperty(this, 'responseText', { writable: true });
        if (mockRequest.timeout) {
          new Promise(resolve => setTimeout(resolve, (mockRequest.timeout || 0) * 1000)).then(() => {
            this?.ontimeout?.({} as any);
          });
          return;
        }
        const originalOnReadyStateChange = this.onreadystatechange;
        this.onreadystatechange = args => {
          try {
            if (mockRequest.response && this.readyState === 4) {
              // We rewrite the response if it is defined
              // @ts-ignore
              this.responseText = mockRequest.response;
              // @ts-ignore
              this.response = mockRequest.response;
            }
            if (mockRequest.statusCode) {
              // We rewrite the status code if it is defined
              // @ts-ignore
              this.status = mockRequest.statusCode;
            }
          } catch (e) {
            console.error(e);
          }
          originalOnReadyStateChange?.apply(this, [args]);
        };
      }
      originalSend.apply(this, [body]);
    };
  } catch (e) {
    console.error(e);
  }
}
