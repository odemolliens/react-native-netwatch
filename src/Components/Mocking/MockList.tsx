import { Button, ScrollView, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import React, { useContext, useEffect, useState } from 'react';
import { ThemeContext } from '../../Theme';
import { NavBar } from '../NavBar';
import { Divider } from 'react-native-paper';
import { clearMockResponses, FILE_PATH, getMockResponses, MockResponse, resetMockResponses } from './utils';
import FeatherIcon from 'react-native-vector-icons/Feather';
import Share from 'react-native-share';
import RNFS from 'react-native-fs';
import Clipboard from '@react-native-clipboard/clipboard';

export function NavbarRightSide(props: {
  mockResponsesCopy: MockResponse[];
  setMockResponsesCopy: (mockResponsesCopy: MockResponse[]) => void;
}) {
  const theme = useContext(ThemeContext);

  return (
    <>
      <TouchableOpacity
        onPress={() => {
          Clipboard.getString().then(responses => {
            if (responses) {
              resetMockResponses(responses);
              props.setMockResponsesCopy(getMockResponses());
            }
          });
        }}
        style={[styles.button, { borderLeftWidth: 0 }]}
      >
        <FeatherIcon name="download" color={theme.textColorOne} size={24} />
      </TouchableOpacity>
      {props.mockResponsesCopy.length > 0 && (
        <TouchableOpacity
          style={[styles.button, { borderLeftWidth: 0 }]}
          onPress={() => {
            RNFS.copyFile(FILE_PATH, `${RNFS.DocumentDirectoryPath}/mocks-preset.json`)
              .then(() => {
                Share.open({
                  title: 'Export Mock preset',
                  url: `${RNFS.DocumentDirectoryPath}/mocks-preset.json`,
                  type: 'text/plain',
                  // excludedActivityTypes: []
                }).then(() => {
                  RNFS.unlink(`${RNFS.DocumentDirectoryPath}/mocks-preset.json`);
                });
              })
              .catch(console.error);
          }}
        >
          <FeatherIcon name="upload" color={theme.textColorOne} size={24} />
        </TouchableOpacity>
      )}
      <Button
        title={'Clear'}
        onPress={() => {
          clearMockResponses();
          props.setMockResponsesCopy([]);
        }}
      />
    </>
  );
}

export function MockList(props: {
  showEdit: (mockResponse: MockResponse, update: boolean) => void;
  onPressBack: () => void;
}) {
  const theme = useContext(ThemeContext);

  const [mockResponsesCopy, setMockResponsesCopy] = useState<MockResponse[]>([]);

  useEffect(() => {
    setMockResponsesCopy(getMockResponses());
  }, []);

  return (
    <View style={styles.container}>
      <NavBar
        icon="x"
        title={'Mock List'}
        onPressBack={props.onPressBack}
        rightComponent={
          <NavbarRightSide mockResponsesCopy={mockResponsesCopy} setMockResponsesCopy={setMockResponsesCopy} />
        }
      />

      <ScrollView style={{ backgroundColor: theme.secondaryColor }} contentContainerStyle={styles.scrollView}>
        {mockResponsesCopy.length === 0 && (
          <Text style={{ color: theme.textColorOne, alignSelf: 'center' }}>No mock created yet</Text>
        )}
        {mockResponsesCopy.length > 0 && (
          <>
            <View style={{ flexDirection: 'row' }}>
              <Text style={[{ flex: 1, color: theme.textColorOne }, styles.title]}>URL</Text>
              <Text style={[styles.title, { color: theme.textColorOne }]}>Method</Text>
              <Text style={[styles.title, { color: theme.textColorOne }]} />
            </View>

            {mockResponsesCopy.map((mock, index) => (
              <TouchableOpacity key={`mock-${index}`} onPress={() => props.showEdit(mock, true)}>
                <View style={{ flexDirection: 'row' }}>
                  <Text style={[{ flex: 1, color: mock.active ? theme.primaryColor : 'grey' }, styles.cell]}>
                    {mock.url}
                  </Text>
                  <Text style={[styles.cell, { color: theme.textColorOne }]}>{mock.method}</Text>
                  <View style={[styles.cell, { minWidth: 30 }]}>
                    <FeatherIcon name="arrow-right" color={theme.textColorOne} size={24} />
                  </View>
                </View>
                <Divider />
              </TouchableOpacity>
            ))}
          </>
        )}
      </ScrollView>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    height: '100%',
    width: '100%',
    top: 0,
    left: 0,
  },
  subheading: {
    paddingHorizontal: 16,
    paddingVertical: 8,
    marginBottom: 10,
  },
  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
  scrollView: {
    paddingTop: 26,
    paddingBottom: 20,
  },

  title: {
    fontWeight: 'bold',
    margin: 10,
    minWidth: 50,
  },

  cell: {
    margin: 10,
    alignSelf: 'flex-start',
    minWidth: 50,
  },
});
