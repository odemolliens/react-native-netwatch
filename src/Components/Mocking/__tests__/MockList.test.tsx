import React, { useState } from 'react';
import { MockList, NavbarRightSide } from '../MockList';
import { shallow, ShallowWrapper } from 'enzyme';
import { clearMockResponses, getMockResponses, resetMockResponses } from '../utils';
import { Button, TouchableOpacity } from 'react-native';
import Clipboard from '@react-native-clipboard/clipboard';
import RNFS from 'react-native-fs';
import Share from 'react-native-share';

jest.mock('react-native-fs', () => ({
  copyFile: jest.fn(),
  unlink: jest.fn(),
}));

jest.mock('react-native-share', () => ({
  open: jest.fn(),
}));

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useEffect: (f: () => void) => f(),
}));

jest.mock('../utils', () => ({
  resetMockResponses: jest.fn(),
  getMockResponses: jest.fn(),
  clearMockResponses: jest.fn(),
}));

describe('Test Mock List Screen', () => {
  let component: ShallowWrapper;
  const setMockResponsesCopy = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (useState as jest.Mock).mockReturnValue([
      [
        {
          url: 'https://www.google.com',
          method: 'GET',
          headers: '',
          response: '',
          statusCode: 0,
          timeout: 0,
          active: true,
          date: 0,
        },
        {
          url: 'https://www.mendixxx.com',
          method: 'GET',
          headers: '',
          response: '',
          statusCode: 0,
          timeout: 0,
          active: false,
          date: 0,
        },
      ],
      jest.fn(),
    ]);
  });

  it('should render properly', () => {
    component = shallow(<MockList showEdit={jest.fn} onPressBack={() => jest.fn()} />);
    expect(component).toMatchSnapshot();
  });

  it('should paste content from clipboard and reset mock list', async () => {
    givenClipboardContent(
      '{"url":"https://www.google.com","method":"GET","headers":"","response":"","statusCode":0,"timeout":0,"active":true,"date":0}',
    );
    givenNavbarRightSide();
    await whenPressingPasteButton();
    thenShouldResetMockResponses();
  });

  it('should not paste anything if clipboard is empty', async () => {
    givenClipboardContent('');
    givenNavbarRightSide();
    await whenPressingPasteButton();
    thenShouldNotResetMockResponses();
  });

  it('should export mocks in a json file properly', async () => {
    givenNavbarRightSide();
    givenFileCreationSucceeds();
    givenFileExportSuccess();
    await whenPressingExportButton();
    thenMocksHasBeenExported();
  });

  it('should delete all mocks when pressing clear button', () => {
    givenNavbarRightSide();
    whenPressingClearButton();
    thenMockAreReset();
  });

  function givenClipboardContent(clipboardContent: string) {
    Clipboard.getString = jest.fn().mockResolvedValue(clipboardContent);
  }

  function givenNavbarRightSide() {
    component = shallow(
      <NavbarRightSide
        mockResponsesCopy={[
          {
            url: 'https://www.mendixxx.com',
            method: 'GET',
            headers: '',
            response: '',
            statusCode: 0,
            timeout: 0,
            active: true,
            date: 0,
          },
        ]}
        setMockResponsesCopy={setMockResponsesCopy}
      />,
    );
  }

  function givenFileCreationSucceeds() {
    (RNFS.copyFile as jest.Mock).mockResolvedValue(true);
  }

  function givenFileExportSuccess() {
    (Share.open as jest.Mock).mockResolvedValue(true);
  }

  function whenPressingExportButton() {
    component.find(TouchableOpacity).at(1).simulate('press');
  }

  function thenMocksHasBeenExported() {
    expect(RNFS.unlink).toHaveBeenCalled();
  }

  async function whenPressingPasteButton() {
    await component.find(TouchableOpacity).at(0).simulate('press');
  }

  function thenShouldResetMockResponses() {
    expect(resetMockResponses).toHaveBeenCalled();
  }

  function whenPressingClearButton() {
    component.find(Button).simulate('press');
  }

  function thenMockAreReset() {
    expect(clearMockResponses).toHaveBeenCalled();
    expect(setMockResponsesCopy).toHaveBeenCalledWith([]);
  }

  function thenShouldNotResetMockResponses() {
    expect(resetMockResponses).not.toHaveBeenCalled();
  }
});
