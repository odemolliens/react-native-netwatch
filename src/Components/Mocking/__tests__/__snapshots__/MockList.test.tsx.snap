// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Mock List Screen should render properly 1`] = `
<View
  style={
    Object {
      "height": "100%",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": "100%",
    }
  }
>
  <NavBar
    icon="x"
    onPressBack={[Function]}
    rightComponent={
      <NavbarRightSide
        mockResponsesCopy={
          Array [
            Object {
              "active": true,
              "date": 0,
              "headers": "",
              "method": "GET",
              "response": "",
              "statusCode": 0,
              "timeout": 0,
              "url": "https://www.google.com",
            },
            Object {
              "active": false,
              "date": 0,
              "headers": "",
              "method": "GET",
              "response": "",
              "statusCode": 0,
              "timeout": 0,
              "url": "https://www.mendixxx.com",
            },
          ]
        }
        setMockResponsesCopy={
          [MockFunction] {
            "calls": Array [
              Array [
                undefined,
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
      />
    }
    title="Mock List"
  />
  <ScrollView
    contentContainerStyle={
      Object {
        "paddingBottom": 20,
        "paddingTop": 26,
      }
    }
    style={
      Object {
        "backgroundColor": "#1F2937",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#F9FAFB",
              "flex": 1,
            },
            Object {
              "fontWeight": "bold",
              "margin": 10,
              "minWidth": 50,
            },
          ]
        }
      >
        URL
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontWeight": "bold",
              "margin": 10,
              "minWidth": 50,
            },
            Object {
              "color": "#F9FAFB",
            },
          ]
        }
      >
        Method
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontWeight": "bold",
              "margin": 10,
              "minWidth": 50,
            },
            Object {
              "color": "#F9FAFB",
            },
          ]
        }
      />
    </View>
    <ForwardRef
      key="mock-0"
      onPress={[Function]}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#06B6D4",
                "flex": 1,
              },
              Object {
                "alignSelf": "flex-start",
                "margin": 10,
                "minWidth": 50,
              },
            ]
          }
        >
          https://www.google.com
        </Text>
        <Text
          style={
            Array [
              Object {
                "alignSelf": "flex-start",
                "margin": 10,
                "minWidth": 50,
              },
              Object {
                "color": "#F9FAFB",
              },
            ]
          }
        >
          GET
        </Text>
        <View
          style={
            Array [
              Object {
                "alignSelf": "flex-start",
                "margin": 10,
                "minWidth": 50,
              },
              Object {
                "minWidth": 30,
              },
            ]
          }
        >
          <Icon
            color="#F9FAFB"
            name="arrow-right"
            size={24}
          />
        </View>
      </View>
      <Component />
    </ForwardRef>
    <ForwardRef
      key="mock-1"
      onPress={[Function]}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "grey",
                "flex": 1,
              },
              Object {
                "alignSelf": "flex-start",
                "margin": 10,
                "minWidth": 50,
              },
            ]
          }
        >
          https://www.mendixxx.com
        </Text>
        <Text
          style={
            Array [
              Object {
                "alignSelf": "flex-start",
                "margin": 10,
                "minWidth": 50,
              },
              Object {
                "color": "#F9FAFB",
              },
            ]
          }
        >
          GET
        </Text>
        <View
          style={
            Array [
              Object {
                "alignSelf": "flex-start",
                "margin": 10,
                "minWidth": 50,
              },
              Object {
                "minWidth": 30,
              },
            ]
          }
        >
          <Icon
            color="#F9FAFB"
            name="arrow-right"
            size={24}
          />
        </View>
      </View>
      <Component />
    </ForwardRef>
  </ScrollView>
</View>
`;
